### PR TITLE
1.14.5 fees UI

### PR DESCRIPTION
**Pull request by rnicoll**

### CODE REVIEW

**Comment by patricklodder** [comment]
Love the humor, question though: how does increasing fee increase confirmation times right now, effectively?

**Comment by michilumin** [comment]
ok. So, checked this out as well and - I like the simplicity and this is what I was aiming for; now, to Patrick's question... that's a good point: If someone were to set a higher fee, would it be mined sooner? With Bitcoin, their blocks are full so there's competition -- ours aren't full. So I think as it stands that slider actually wouldn't affect transaction priority. (Of course some day down the line, which we hope doesn't happen - it could.)  So maybe having the slider there is OK, but perhaps we can add a note that says "This slider will have no effect on speed unless network conditions"... something (i don't know the right phrasing right now..)

Technically, though, the review is a thumbs up...

**Review by michilumin — COMMENTED** [review]
Conceptually: good, undecided on the phrasing (I was on board for adding that as like an easter egg...) -   though that may be tougher to do of course. (or not?)   In any case, this does away with the 'multipliers' i was using which is more straightforward, using a case statement.

**Comment by patricklodder** [comment]
So, linking fee height to confirmation speed CAN be a thing right now:

- If you pay 1 DOGE per kb including rounding up, then you can be positive that ALL nodes relay your tx and that it gets mined quickly.
- If you don't then you will only be relayed by 1.14.4+ nodes and only be mined by those miners that have upgraded past 1.14.3

However, I do not see that reflected in the implementation, so I think the speed thing is a bit misleading, at least for the current network.

**Comment by rnicoll** [comment]
It doesn't actually change confirmation time currently, I've put different fee values in for now but we could fix the number entirely, remove the slider. This is here because we wanted to have the slider do something, and it moves the discussion forward.

File: src/dogecoin-fees.cpp [hunk 1]
```diff
@@ -18,6 +18,24 @@
...
+CFeeRate GetDogecoinFeeRate(int speed)
+{
+    switch(speed)
+    {
+    case SUCH_FAST:
+        return CFeeRate(::minWalletTxFeeRate.GetFeePerK() * 10);
+    case MUCH_SPEED:
+        return CFeeRate(::minWalletTxFeeRate.GetFeePerK() * 5);
+    case WOW:
+        return CFeeRate(::minWalletTxFeeRate.GetFeePerK() * 2.5);
```
Comment (rnicoll) [inline src/dogecoin-fees.cpp:31]:
```suggestion
        return CFeeRate(::minWalletTxFeeRate.GetFeePerK() * 100);
    case MUCH_SPEED:
        return CFeeRate(::minWalletTxFeeRate.GetFeePerK() * 50);
    case WOW:
        return CFeeRate(::minWalletTxFeeRate.GetFeePerK() * 25);
```

**Comment by rnicoll** [comment]
I've updated the placeholder fee values, but would note I don't know if we're targetting four options, so need to decide what the UI options as a precursor to finalizing the numbers.

**Comment by patricklodder** [comment]
So let's take a step back. The following situations can occur:

1. Connected to older peers, lowest fee filter is set at 1 DOGE/kb **_plus rounding_** or, say, most miners haven't upgraded yet - _this means the maximum fee rate is 5.21 DOGE/kb (as the smallest tx the wallet makes would be 192kb)_
2. Miners do upgrade but don't lower the fees - _this would put the fees at 100x recommended fee rate_
3. Miners do upgrade but decide that they only appreciate `fees/10`, not `fees/100` -  _this would put the fees at 10x recommended fee rate_
4. Blocks are full, people need increments to get some sort of priority - _2x, 5x would probably do, or alternatively 2x 4x 8x_

So the slider would then be:

```
0.01    0.02    0.05    0.1     1.0   5.21
|--------|-------|-------|-------|-------|
```

**Comment by rnicoll** [comment]
I think I'm inclined to push the slider back to 1.14.6, remove it entirely from 1.14.5 so we have a functioning solution with lower fees soon, and can examine adoption before deciding on this. What do others think?

**Comment by ReverseControl** [comment]
@rnicoll  I agree.

**Comment by patricklodder** [comment]
The network currently is ~22.5% 1.14.4 and adoption is growing very slowly the past week. At 22.5%, this means 77.5%^8 = 13% of shibes will not connect to a 1.14.4 right now so it's too early to release. There is time to get this done and the slider is a user-friendly way of letting people determine their fees. So why not take the time to do it right? I don't see ANY pressure to release other than people wanting to take credit for the release.

**Comment by ReverseControl** [comment]
@patricklodder That is a fair point. There are still quite a few things left on the dock that are high priority to get done  -- mostly security related stuff including #2459. If by the time we  are done with all those this is taken care of then all is fine. If we finish everything else first, then I would support releasing and delaying this one.

**Comment by patricklodder** [comment]
> There are still quite a few things left on the dock [..]

Agree, but it's mostly important to prevent another fee disaster, as where last November, about 11% of transactions got stuck in the mempools, it would be a fantastic CF if devs would create the same problem again, by rushing a fee recommendation before the network is ready, and before there are controls in place to allow a resolution for transactions stuck and even more so since the recommendation PR got rushed by a private review deal between "core devs" while I had a review in progress on it, of which the devs that rushed it were aware, and as a result, the current fee logic is not fully consistent yet, and readability from a "reference implementation" point of view is terrifying. So we have time, and we need time: for this UI fix, for a consistent set of fees, and for the network to adapt to have a more reliably lenient mempool.

I was also hoping to get these UI changes in so that we can fixate translations, because that's a real mess, and then maybe we can get some internationalization fixed **before** release rather than just English and German, both of which are less important than for example Asian translations. I was also hoping to be able to do some build system and dependency upgrades while we still have time, but all this is getting endangered from an artificial release pressure. I understand people have a need for showing the world how awesome their new org is by doing press releases, but not at the cost of another round of transactions getting stuck for no reason other than developer vanity.

***PS:***

> [..] high priority to get done -- mostly security related stuff including #2459.

#2459 cannot be high priority because in the field, 1.14 nodes do not use `cmpctblock` for announcements and on top, https://github.com/dogecoin/dogecoin/pull/2562#issuecomment-917395863 from today claims that it is broken (?!?) If anything regarding that is high priority it would probably - especially since today's claim - be to re-enable the test and make sure that whenever 1.21 nodes are going to imply they need `cmpctblock` messages, 1.14 nodes are not going to split the network up. The more 1.14 versions we can get out there that actually are tested to run the protocol they are advertising to support (meaning: `cmpctblock` by announcing version 70015) is going to help MUCH more than disconnecting a peer in an edge scenario for a protocol that requires leniency on pre-validation announces, which is guaranteed by that same version announcement. So if we want to bring that forward, we have to fix `p2p-compactblocks.py` first, then assess whether we need to do something with the protocol implementation per the claim in above mentioned issue comment, and only then will we be able to do something real with #2459.

**Comment by ReverseControl** [comment]
@patricklodder I was referring to enabling enabling testing as you mentioned here, https://github.com/dogecoin/dogecoin/pull/2459#issuecomment-902982023. So, I think we both agree enabling that test and updating it is a priority. The resolution of the issue itself is another matter.

> The more 1.14 versions we can get out there that actually are tested to run the protocol they are advertising to support (meaning: cmpctblock by announcing version 70015) is going to help MUCH more than disconnecting a peer in an edge scenario for a protocol that requires leniency on pre-validation announces, which is guaranteed by that same version announcement.

I agree. Given we moved #2459 from the last release, and that it was a quick fix to #2383, I wanted to get feedback on it. The work I did initially suggested fixing the original issue might be a lot of work, https://github.com/dogecoin/dogecoin/issues/2283#issuecomment-900708165. But, as you point out, it is better to get the testing working again and enforce the implementation later after assessing the situation then.

>So if we want to bring that forward, we have to fix p2p-compactblocks.py first, then assess whether we need to do something with the protocol implementation per the claim in above mentioned issue comment, 

We are in agreement.

**Comment by rnicoll** [comment]
Closing this, but keeping the branch in case we want to re-visit.

**Comment by rnicoll** [comment]
As said elsewhere - this was a UI mock up, I hadn't expected it to be reviewed for merging, so much as demonstrate what the UI could look like.

If however people do like the slider and it's just a matter of the fee boundaries, happy to re-open this. There are three options for the UI that I am aware of:

* Change the slider to provide Dogecoin-friendly fee values (some variation of this).
  * Conscious that criticising my own PR looks strange, but having made this, I am not sure having values between 1.0 and 0.01 is useful.
* Remove the slider entirely, and change the default action to be 0.01 DOGE fees.
  * Aware there are concerns of reliability, but initial tests suggest low fee transactions are mined consistently within minutes. I'll run more tests on main.
* Replace the slider and custom with a pair of radio boxes, with 1.0 and 0.01 fees. Have not yet mocked up this option.
  * Removes options the user shouldn't be using currently.
  * More invasive UI change, makes it harder to re-introduce variable fees later.

I'll update this with Patrick's suggested boundaries and re-open it tonight. If we want a mock-up of the last option do let me know.

**Comment by rnicoll** [comment]
Today I learned if you force-push a closed branch, it won't let you re-open. I'll make a new PR. Sorry about that.

**Pull request closed without merging**